model_signing/signing/in_toto.py

```diff
@@ -0,0 +1,256 @@
+    present in the resource descriptor (i.e., sha256).
+
+    See also https://github.com/sigstore/sigstore-python/issues/1018.
+    """
+
+    predicate_type: Final[str] = "https://model_signing/Digest/v0.1"
+
+    def __init__(self, *, digest_hex: str, digest_algorithm: str):
+        """Builds an instance of this in-toto payload.
+
+        Don't call this directly in production. Use `from_manifest()` instead.
+
+        Args:
+            digest_hex: the hexadecimal, human readable, digest of the subject.
+            digest_algorithm: the algorithm used to compute the digest.
+        """
+        digest = {"sha256": digest_hex}
+        descriptor = statement.ResourceDescriptor(digest=digest).pb
+
+        self.statement = statement.Statement(
+            subjects=[descriptor],
+            predicate_type=self.predicate_type,
+            predicate={"actual_hash_algorithm": digest_algorithm},
+        )
+
+    @classmethod
+    @override
+    def from_manifest(cls, manifest: manifest_module.Manifest) -> Self:
+        """Converts a manifest to the signing payload used for signing.
+
+        The manifest must be a `DigestManifest` instance.
+
+        Args:
+            manifest: the manifest to convert to signing payload.
+
+        Returns:
+            An instance of `SingleDigestIntotoPayload`.
+
+        Raises:
+            TypeError: If the manifest is not `DigestManifest`.
+        """
+        if not isinstance(manifest, manifest_module.DigestManifest):
+            raise TypeError("Only DigestManifest is supported")
+
+        # guaranteed to have exactly one item
+        subject = list(manifest.resource_descriptors())[0]
+        digest = subject.digest
+        return cls(
+            digest_hex=digest.digest_hex,
+            digest_algorithm=digest.algorithm,
+        )
+
+
+def _convert_descriptors_to_hashed_statement(
+    manifest: manifest_module.Manifest,
+    *,
+    predicate_type: str,
+    predicate_top_level_name: str,
+):
+    """Converts manifest descriptors to an in-toto statement with payload.
+
+    Args:
+        manifest: The manifest to extract the descriptors from. Assumed valid.
+        predicate_type: The predicate_type to use in the in-toto statement.
+        predicate_top_level_name: Name to use in the payload for the array of
+          the subjects.
+    """
+    hasher = memory.SHA256()
+    subjects = []
+    for descriptor in manifest.resource_descriptors():
+        hasher.update(descriptor.digest.digest_value)
+        subjects.append({
+            "name": descriptor.identifier,
+            "digest": descriptor.digest.digest_hex,
+            "algorithm": descriptor.digest.algorithm,
+        })
+
+    digest = {"sha256": hasher.compute().digest_hex}
+    descriptor = statement.ResourceDescriptor(digest=digest).pb
+
+    return statement.Statement(
+        subjects=[descriptor],
+        predicate_type=predicate_type,
+        predicate={predicate_top_level_name: subjects},
+    )
```

Comment:
Shouldn't the subject be a list of the file hashes?

It feels a bit like misusing the Statement format when the digests are put into the predicate.

Reply:
Doing that in #266. But in #111 there was a discussion that the semantics of verification on subjects is not working properly for models, so we put as subject a single digest for the entire model and then the actual hashes in payload to reconstruct as needed.

Reply:
I don't really see the verification argument. It applies to every format you roll with, if there is another implementation there is always the risk of doing it wrong/different.

Reply:
I just implemented all manifest ideas discussed in #111 so we can experiment with them and see which one is more ergonomic, better suited for our needs, less prone to failures

Reply:
Also, this specific PR is what's being discussed in https://github.com/sigstore/model-transparency/issues/248#issuecomment-2248298175